### PR TITLE
fix: fix shadowstack pass discard debuginfo

### PR DESCRIPTION
**Pull request by HerrCai0907**
In shadowstack pass, if some function need to add temp local, It need to be removed and re-added. It will discard all debug info because debug info bind with functionref.

This PR aims at re-assign the debug info after shadowstack pass

- [x] I've read the contributing guidelines
- [x] I've added my name and email to the NOTICE file

### CODE REVIEW

**Comment by dcodeIO** [comment]
Would it perhaps be preferable to add a `replaceFunction` helper of sorts that can be called instead of removing the function and adding a new one? I guess then we'd not need to remember intermediate state.

**Comment by MaxGraey** [comment]
Yeah, also prefer update 
`updateFunction(funcRef: FunctionRef)` 
to 
`updateFunction(funcRef: FunctionRef, preserveDebugInfo: bool)`

**Comment by HerrCai0907** [comment]
> Would it perhaps be preferable to add a `replaceFunction(..., preserveDebugInfo = true)` helper of sorts that can be called instead of removing the function and adding a new one? I guess then we'd not need to remember intermediate state.

You mean in add this API in binaryen? I also agree it is more elegant

File: src/passes/shadowstack.ts [hunk 1]
```diff
@@ -477,7 +478,9 @@ export class ShadowStackPass extends Pass {
     let moduleRef = this.module.ref;
     _BinaryenRemoveFunction(moduleRef, name);
     let cArr = allocPtrArray(vars);
-    _BinaryenAddFunction(moduleRef, name, params, results, cArr, vars.length, body);
+    let newFuncRef = _BinaryenAddFunction(moduleRef, name, params, results, cArr, vars.length, body);
+    let func = Function.searchFunctionByRef(this.compiler, newFuncRef);
+    if (func) func.addDebugInfo(this.module, newFuncRef);
```
Comment (MaxGraey) [inline src/passes/shadowstack.ts:483]:
```suggestion
    if (this.options.sourceMap || this.options.debugInfo) {
      let func = Function.searchFunctionByRef(this.compiler, newFuncRef);
      if (func) func.addDebugInfo(this.module, newFuncRef);
    }
```
For skipping unnecessary search routine when we don't actually need this.

**Review by MaxGraey — APPROVED** [review]

File: src/passes/shadowstack.ts [hunk 2]
```diff
@@ -477,7 +478,11 @@ export class ShadowStackPass extends Pass {
     let moduleRef = this.module.ref;
     _BinaryenRemoveFunction(moduleRef, name);
     let cArr = allocPtrArray(vars);
-    _BinaryenAddFunction(moduleRef, name, params, results, cArr, vars.length, body);
+    let newFuncRef = _BinaryenAddFunction(moduleRef, name, params, results, cArr, vars.length, body);
+    if (this.options.sourceMap || this.options.debugInfo) {
+      let func = Function.searchFunctionByRef(this.compiler, newFuncRef);
```
Comment (dcodeIO) [inline src/passes/shadowstack.ts:483]: The shadow stack pass has a reference to the `Program` here via the compiler. With this information, I think `Function.searchFunctionByRef` could be simplified to `program.searchFunctionByRef`?

**Pull request merged**